File: tools/verify_canonicalization_vectors.mjs
#!/usr/bin/env node
// Independent Node.js verifier for CPAS RFC 8785/JCS and digest vectors.

import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { resolve } from "node:path";

const JCS_PROFILE = "rfc8785-jcs-v1";
const LEGACY_PROFILE = "cpas-canonical-json-v1";
const LEGACY_DIGEST_PROFILE = "cpas-sha256-direct-v1";
const DIGEST_MAGIC = Buffer.from("CPAS-DIGEST-V2\0", "ascii");
const DEFAULT_VECTORS = resolve(
  "compliance-tests/canonicalization/cpas-canonicalization-v1.json",
);

class DuplicateKeyError extends Error {}
class NonFiniteNumberError extends Error {}
class IntegerDomainError extends Error {}
class InvalidUnicodeError extends Error {}

function assertScalarUnicode(value) {
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) {
        throw new InvalidUnicodeError("unpaired high surrogate");
      }
      index += 1;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      throw new InvalidUnicodeError("unpaired low surrogate");
    }
  }
}

function canonicalize(value) {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new NonFiniteNumberError("non-finite number");
    }
    return JSON.stringify(value);
  }
  if (typeof value === "string") {
    assertScalarUnicode(value);
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => canonicalize(item)).join(",")}]`;
  }
  if (typeof value === "object") {
    const entries = Object.keys(value)
      .sort()
      .map((key) => {
        assertScalarUnicode(key);
        return `${JSON.stringify(key)}:${canonicalize(value[key])}`;
      });
    return `{${entries.join(",")}}`;
  }
  throw new TypeError(`unsupported JSON value type: ${typeof value}`);
}

// JSON.parse discards duplicate object members. This small recursive parser
// preserves the parser-boundary guarantee required by the negative vectors.
function parseStrictJson(text) {
  let index = 0;
  const whitespace = /[\u0009\u000a\u000d\u0020]/;

  function skipWhitespace() {
    while (index < text.length && whitespace.test(text[index])) index += 1;
  }

  function parseString() {
    const start = index;
    if (text[index] !== '"') throw new SyntaxError("expected string");
    index += 1;
    while (index < text.length) {
      const character = text[index];
      if (character === "\\") {
        index += 2;
        continue;
      }
      if (character === '"') {
        index += 1;
        return JSON.parse(text.slice(start, index));
      }
      if (character.charCodeAt(0) < 0x20) {
        throw new SyntaxError("unescaped control character");
      }
      index += 1;
    }
    throw new SyntaxError("unterminated string");
  }

  function parseArray() {
    const result = [];
    index += 1;
    skipWhitespace();
    if (text[index] === "]") {
      index += 1;
      return result;
    }
    while (true) {
      result.push(parseValue());
      skipWhitespace();
      if (text[index] === "]") {
        index += 1;
        return result;
      }
      if (text[index] !== ",") throw new SyntaxError("expected array comma");
      index += 1;
      skipWhitespace();
    }
  }

  function parseObject() {
    const result = {};
    const keys = new Set();
    index += 1;
    skipWhitespace();
    if (text[index] === "}") {
      index += 1;
      return result;
    }
    while (true) {
      const key = parseString();
      if (keys.has(key)) throw new DuplicateKeyError(`duplicate key: ${key}`);
      keys.add(key);
      skipWhitespace();
      if (text[index] !== ":") throw new SyntaxError("expected object colon");
      index += 1;
      skipWhitespace();
      result[key] = parseValue();
      skipWhitespace();
      if (text[index] === "}") {
        index += 1;
        return result;
      }
      if (text[index] !== ",") throw new SyntaxError("expected object comma");
      index += 1;
      skipWhitespace();
    }
  }

  function parseValue() {
    skipWhitespace();
    if (
      text.startsWith("NaN", index) ||
      text.startsWith("Infinity", index) ||
      text.startsWith("-Infinity", index)
    ) {
      throw new NonFiniteNumberError("non-finite number token");
    }
    if (text[index] === '"') return parseString();
    if (text[index] === "[") return parseArray();
    if (text[index] === "{") return parseObject();
    for (const [token, value] of [
      ["true", true],
      ["false", false],
      ["null", null],
    ]) {
      if (text.startsWith(token, index)) {
        index += token.length;
        return value;
      }
    }
    const number = text.slice(index).match(/^-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?/);
    if (!number) throw new SyntaxError("expected JSON value");
    index += number[0].length;
    const parsed = Number(number[0]);
    if (!/[.eE]/.test(number[0]) && !Number.isSafeInteger(parsed)) {
      throw new IntegerDomainError("integer token outside the safe JSON domain");
    }
    return parsed;
  }

  const result = parseValue();
  skipWhitespace();
  if (index !== text.length) throw new SyntaxError("trailing JSON content");
  return result;
}

function sha256(bytes) {
  return `sha256:${createHash("sha256").update(bytes).digest("hex")}`;
}

function profiledDigest(value, profile) {
  const canonical = Buffer.from(canonicalize(value), "utf8");
  const preimage = Buffer.concat([
    DIGEST_MAGIC,
    Buffer.from(profile, "ascii"),
    Buffer.from([0]),
    Buffer.from(JCS_PROFILE, "ascii"),
    Buffer.from([0]),
    canonical,
  ]);
  return sha256(preimage);
}

function classifyNegative(input) {
  try {
    canonicalize(parseStrictJson(input));
  } catch (error) {
    if (error instanceof DuplicateKeyError) return "duplicate_key";
    if (error instanceof NonFiniteNumberError) return "non_finite_number";
    if (error instanceof IntegerDomainError) return "integer_domain";
    if (error instanceof InvalidUnicodeError) return "invalid_unicode";
    return "json_parse";
  }
  return null;
}

function fail(message) {
  throw new Error(message);
}

function verify(path) {
  const vectors = parseStrictJson(readFileSync(path, "utf8"));
  if (vectors.canonicalization !== JCS_PROFILE) fail("canonicalization marker differs");
  if (vectors.digest_frame.magic_hex !== DIGEST_MAGIC.toString("hex")) {
    fail("digest frame magic differs");
  }
  let checked = 0;
  for (const vector of vectors.positive) {
    const canonical = Buffer.from(canonicalize(vector.value), "utf8");
    if (canonical.toString("hex") !== vector.canonical_hex) {
      fail(`${vector.id}: canonical bytes differ`);
    }
    if (canonical.length !== vector.canonical_length) {
      fail(`${vector.id}: canonical length differs`);
    }
    for (const [profile, expected] of Object.entries(vector.digests)) {
      if (profiledDigest(vector.value, profile) !== expected) {
        fail(`${vector.id}: digest differs for ${profile}`);
      }
      checked += 1;
    }
  }

  const legacy = vectors.legacy;
  if (legacy.canonicalization !== LEGACY_PROFILE) fail("legacy marker differs");
  if (legacy.digest_profile !== LEGACY_DIGEST_PROFILE) {
    fail("legacy digest profile differs");
  }
  const legacyBytes = Buffer.from(canonicalize(legacy.value), "utf8");
  if (legacyBytes.toString("hex") !== legacy.canonical_hex) {
    fail("legacy canonical bytes differ");
  }
  if (sha256(legacyBytes) !== legacy.digest) fail("legacy digest differs");
  checked += 1;

  for (const vector of vectors.negative) {
    const actual = classifyNegative(vector.input_json);
    if (actual !== vector.expected_error) {
      fail(`${vector.id}: expected ${vector.expected_error}, observed ${actual}`);
    }
    checked += 1;
  }
  return checked;
}

try {
  const path = resolve(process.argv[2] ?? DEFAULT_VECTORS);
  const checked = verify(path);
  console.log(`canonicalization vector verification passed: ${checked} checks (Node.js)`);
} catch (error) {
  console.error(`canonicalization vector verification failed: ${error.message}`);
  process.exitCode = 1;
}

File: tools/verify_canonicalization_vectors.py
#!/usr/bin/env python3
"""Verify CPAS canonicalization and domain-separated digest vectors."""

from __future__ import annotations

import argparse
import sys
from pathlib import Path
from typing import Any, Mapping

import rfc8785

REPOSITORY_ROOT = Path(__file__).resolve().parents[1]
if str(REPOSITORY_ROOT) not in sys.path:
    sys.path.insert(0, str(REPOSITORY_ROOT))

from cpas.provenance import (  # noqa: E402
    DIGEST_FRAME_MAGIC,
    JCS_CANONICALIZATION,
    LEGACY_CANONICALIZATION,
    LEGACY_DIGEST_PROFILE,
    DuplicateKeyError,
    canonicalize_json,
    load_json,
    loads_json,
    profiled_digest,
)

DEFAULT_VECTORS = (
    REPOSITORY_ROOT
    / "compliance-tests"
    / "canonicalization"
    / "cpas-canonicalization-v1.json"
)


class VectorFailure(RuntimeError):
    """Raised when a normative vector does not reproduce exactly."""


def _mapping(value: Any, *, owner: str) -> Mapping[str, Any]:
    if not isinstance(value, Mapping):
        raise VectorFailure(f"{owner}: expected an object")
    return value


def classify_negative_vector(input_json: str) -> str | None:
    """Return the normative failure class, or ``None`` if input is accepted."""

    try:
        value = loads_json(input_json)
    except DuplicateKeyError:
        return "duplicate_key"
    except ValueError as exc:
        if "non-finite JSON number" in str(exc):
            return "non_finite_number"
        return "json_parse"
    try:
        canonicalize_json(value, profile=JCS_CANONICALIZATION)
    except rfc8785.IntegerDomainError:
        return "integer_domain"
    except (rfc8785.CanonicalizationError, UnicodeEncodeError):
        return "invalid_unicode"
    return None


def verify_vectors(path: str | Path = DEFAULT_VECTORS) -> int:
    vectors = _mapping(load_json(path), owner="vector document")
    canonicalization = str(vectors.get("canonicalization"))
    if canonicalization != JCS_CANONICALIZATION:
        raise VectorFailure(f"unexpected canonicalization: {canonicalization}")

    frame = _mapping(vectors.get("digest_frame"), owner="digest_frame")
    if frame.get("magic_hex") != DIGEST_FRAME_MAGIC.hex():
        raise VectorFailure("digest frame magic does not match the implementation")

    checked = 0
    for raw_vector in vectors.get("positive", []):
        vector = _mapping(raw_vector, owner="positive vector")
        vector_id = str(vector.get("id", "<unnamed>"))
        canonical = canonicalize_json(vector.get("value"), profile=canonicalization)
        if canonical.hex() != vector.get("canonical_hex"):
            raise VectorFailure(f"{vector_id}: canonical bytes differ")
        if len(canonical) != vector.get("canonical_length"):
            raise VectorFailure(f"{vector_id}: canonical length differs")
        digests = _mapping(vector.get("digests"), owner=f"{vector_id} digests")
        for profile, expected in digests.items():
            actual = profiled_digest(
                vector.get("value"),
                canonicalization=canonicalization,
                digest_profile=str(profile),
            )
            if actual != expected:
                raise VectorFailure(f"{vector_id}: digest differs for {profile}")
            checked += 1

    legacy = _mapping(vectors.get("legacy"), owner="legacy vector")
    if legacy.get("canonicalization") != LEGACY_CANONICALIZATION:
        raise VectorFailure("legacy vector canonicalization marker differs")
    if legacy.get("digest_profile") != LEGACY_DIGEST_PROFILE:
        raise VectorFailure("legacy vector digest profile marker differs")
    legacy_value = legacy.get("value")
    legacy_bytes = canonicalize_json(
        legacy_value,
        profile=LEGACY_CANONICALIZATION,
    )
    if legacy_bytes.hex() != legacy.get("canonical_hex"):
        raise VectorFailure("legacy canonical bytes differ")
    legacy_digest = profiled_digest(
        legacy_value,
        canonicalization=LEGACY_CANONICALIZATION,
        digest_profile=LEGACY_DIGEST_PROFILE,
    )
    if legacy_digest != legacy.get("digest"):
        raise VectorFailure("legacy direct digest differs")
    checked += 1

    for raw_vector in vectors.get("negative", []):
        vector = _mapping(raw_vector, owner="negative vector")
        vector_id = str(vector.get("id", "<unnamed>"))
        actual = classify_negative_vector(str(vector.get("input_json", "")))
        if actual != vector.get("expected_error"):
            raise VectorFailure(
                f"{vector_id}: expected {vector.get('expected_error')}, observed {actual}"
            )
        checked += 1
    return checked


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("path", nargs="?", type=Path, default=DEFAULT_VECTORS)
    args = parser.parse_args(argv)
    try:
        checked = verify_vectors(args.path)
    except Exception as exc:
        print(f"canonicalization vector verification failed: {exc}", file=sys.stderr)
        return 1
    print(f"canonicalization vector verification passed: {checked} checks")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
